### DISCUSSION TITLE
it's possible to sign documents using certificates from windows store?

### DISCUSSION THREAD

**Discussion by Fulea-Andrei** [MatthiasValvekens/pyHanko#137] [Q&A]
hello,

I'm looking for a library to sign PDF files with an instance of the certificate directly from the windows store ( certificates with token usb or others), not from an exported file on the local version. Your library can do this?

I'm using wincertstore.CertSystemStore to get certificates available in python.

thank you and congrats on your work.

**Comment by MatthiasValvekens** [comment]
Hi @Fulea-Andrei,

Yes and no. PyHanko certainly doesn't do that out of the box, but you should be able to relatively easily extend pyHanko to support this use case by [defining your own `Signer` implementation](https://pyhanko.readthedocs.io/en/latest/lib-guide/signing.html#extending-signer) that uses the Windows Crypto APIs. I'm not sure if there exist Python bindings for that, but if not, I'm sure it can be done by slotting in a Cython/ctypes/... module to handle the native parts.

Now, I can't write that integration myself, since I don't have a Windows development machine on hand. I'd also be a little hesitant to merge platform-dependent code into pyHanko for similar reasons, but if you find a way to get this working in a more or less compact manner, I'd be happy to accept a documentation PR with some sample code ([see e.g. this example](https://pyhanko.readthedocs.io/en/latest/lib-guide/adv-examples.html#async-aws-kms)).

**Reply by Fulea-Andrei** [reply]
hello,
thank you for your response. I want to extend a little bit this discussion. 
I want to mention that I'm a beginner in all that means digital signatures, and maybe some expressions will not have logic, please keep it in mind. :) 
So,
My task is to develop an application that selects the certifications stored in windows store certification ( based on USB token most of them) and by selecting one certificate, to sign a specific PDF. For the task of viewing certification from win store, I'm using this blog : 
[https://blog.darrenjrobinson.com/accessing-the-windows-certificate-store-using-python/](url)

I found some libraries such as endesive or your, but I'm seeing in the code pieces that the signature is made with exported version of that certificate from win store, with a file with extension ".p12".
Using this code the certificate is loaded and fed the signature method ( [https://github.com/m32/endesive/blob/master/examples/document_signing/pdf-sign-cms-existing-field.py](url) )

 `  with open("certificate.p12", "rb") as fp:
        p12 = pkcs12.load_key_and_certificates(
            fp.read(), b"1234", backends.default_backend()
        )` .

Basically, I have the code for selecting certifications, probably that library wincertstore.CertSystemStore is wrapping the windows API. 
I simply don't know how to pass the variable with an instance of a certificate from the windows store compared with an instance of variable loaded by method from above ( pkcs12.load_key_and_certificates() ), because they have a different data type, I print some outputs of the data types but I don't understand so much...

I hope that I was clear ...
Thank you a lot for your time and help.

**Reply by MatthiasValvekens** [reply]
So, as I said, I'm not all that familiar with Windows's crypto APIs, but there are a number of things I can say based on general principles.

> My task is to develop an application that selects the certifications stored in windows store certification ( based on USB token most of them) and by selecting one certificate, to sign a specific PDF. For the task of viewing certification from win store, I'm using this blog :
[https://blog.darrenjrobinson.com/accessing-the-windows-certificate-store-using-python/](https://blog.darrenjrobinson.com/accessing-the-windows-certificate-store-using-python/)

Sure, but accessing the certificate store is only half the battle. You can't sign using only a certificate. As you pointed out yourself, the actual private keys (which you need to produce the signature) likely reside in USB tokens or require some kind of trusted platform module to use. In other words, the keys can't be extracted, so you need the token to perform (part of) the signing operation for you. I doubt `wincertstore` can do that.

> I found some libraries such as endesive or your, but I'm seeing in the code pieces that the signature is made with exported version of that certificate from win store, with a file with extension ".p12".

Yeah, the examples use in-memory keys because those are easy to do demonstrations with. In practice, you'll usually have to deal with keys that you _can't_ access directly. PyHanko was designed to support that use case, but since the details depend on the type of token/signing service/... you're using to handle the cryptographic signing operation, it doesn't come with "batteries included" for all possible scenarios.

There is a [pretty generic PKCS#11 signer available](https://pyhanko.readthedocs.io/en/latest/api-docs/pyhanko.sign.pkcs11.html#pyhanko.sign.pkcs11.PKCS11Signer), though. So, if your USB tokens speak PKCS#11, you might be able to proceed that way. That said, maybe they're the kind of token that can only be addressed through Windows-specific APIs, in which case you'll have to supply your own `Signer` implementation (see my previous comment).

> Basically, I have the code for selecting certifications, probably that library wincertstore.CertSystemStore is wrapping the windows API. I simply don't know how to pass the variable with an instance of a certificate from the windows store compared with an instance of variable loaded by method from above ( pkcs12.load_key_and_certificates() ), because they have a different data type, I print some outputs of the data types but I don't understand so much...

I'm sure there's a way to extract the actual certificate data from your call to `wincertstore.CertSystemStore`. The real problem is that the certificate by itself is useless for _creating_ new signatures (after all, certificates are considered public information; they're used to _validate_ signatures, not to create them). You'll need to figure out a way to interact with the signing hardware first :).


TL;DR: You need to tell pyHanko how to feed the data to be signed to your signing hardware. That's the main problem. Once you manage to solve that, retrieving the relevant certificates from the Windows certificate store in a form that pyHanko understands will probably be a lot easier. You can then leave all the PDF-level operations to pyHanko.

**Reply by Fulea-Andrei** [reply]
your explanation is really great, I will read it a few times, I'm going to learn some new things. Thank you for the time to respond to me! have a nice day!

**Reply by Kugeleis** [reply]
Is there progress on this? I'm trying similar stuff.
